File: ruby/day2/day2.rb
p1 = 0
p2 = 0
File.read('input').lines do |line|
  range, letter, password = line.strip.split(' ')
  low, high = range.split('-').map(&:to_i)
  letter = letter[0]

  if password.count(letter) >= low && password.count(letter) <= high
    p1 += 1
  end

  if [password[low-1], password[high-1]].count(letter) == 1
    p2 += 1
  end
end

p p1
p p2

File: haskell/day2/day2.hs
import Data.List.Split

data Entry = Entry Int Int Char String deriving (Show) -- low, high, letter, password

parseEntry :: String -> Entry
parseEntry input = Entry low high (head letter) password
  where [range, letter, password] = splitOn " " input
        [low, high] = map (read :: String -> Int) . splitOn "-" $ range

countChars :: Char -> String -> Int
countChars c = length . filter (== c)

part1 :: [Entry] -> Int
part1 = length . filter validate1

validate1 :: Entry -> Bool
validate1 (Entry low high letter password) = count >= low && count <= high
  where count = countChars letter password

part2 :: [Entry] -> Int
part2 = length . filter validate2

validate2 :: Entry -> Bool
validate2 (Entry low high letter password) = 1 == (countChars letter [x, y])
  where x = password !! (low - 1)
        y = password !! (high - 1)

main = do
    raw_input <- readFile "input"
    let input = map parseEntry . lines $ raw_input

    print $ part1 input
    print $ part2 input
